python/lsst/obs/base/gen2to3/translators.py

```diff
@@ -214,17 +215,49 @@ def extract(self, gen2id: dict, skyMap: Optional[BaseSkyMap], skyMapName: Option
 class CalibKeyHandler(KeyHandler):
     """A KeyHandler for master calibration datasets.
     """
+    __slots__ = ("ccdKey",)
 
-    def __init__(self):
+    def __init__(self, ccdKey="ccd"):
+        self.ccdKey = ccdKey
         super().__init__("calibration_label")
 
-    __slots__ = ()
-
     def extract(self, gen2id: dict, skyMap: Optional[BaseSkyMap], skyMapName: Optional[str],
                 datasetTypeName: str) -> Any:
         # Docstring inherited from KeyHandler.extract.
         return makeCalibrationLabel(datasetTypeName, gen2id["calibDate"],
-                                    ccd=gen2id.get("ccd"), filter=gen2id.get("filter"))
+                                    ccd=gen2id.get(self.ccdKey), filter=gen2id.get("filter"))
+
+
+class AbstractFilterKeyHandler(KeyHandler):
+    """KeyHandler for filter keys that should be mapped to AbstractFilters.
+    """
+
+    __slots__ = ("_map",)
+
+    def __init__(self, filterDefinitions):
+        super().__init__("abstract_filter")
+        self._map = {d.physical_filter: d.abstract_filter for d in filterDefinitions
+                     if d.physical_filter is not None}
+
+    def extract(self, gen2id, *args, **kwargs):
+        physical = gen2id["filter"]
+        return self._map.get(physical, physical)
+
+
+class PhysicalFilterKeyHandler(KeyHandler):
+    """KeyHandler for filter keys that should be mapped to AbstractFilters.
+    """
+
+    __slots__ = ("_map",)
+
+    def __init__(self, filterDefinitions):
+        super().__init__("physical_filter")
+        self._map = {d.abstract_filter: d.physical_filter for d in filterDefinitions
```

Comment:
By definition multiple physical filters can map to the same abstract filter but the above code doesn't take that into account (currently afwFilter also breaks of course but that will get fixed).

Reply:
I don't know if this translator rule will apply to HSC, but if it does, `i -> HSC-I, HSC-I2` is an example of that being a one-to-many relation, even within a single camera.  And unfortunately getting it wrong won't trigger any CI test failures, because data taken with I2 (or R2) went public after our test data repositories were put together.

Reply:
These two translators have to be set up by a given obs package, so would only apply if the situation is valid for that obs package (e.g. DECam uses an `abstract` name in its gen2 calibRegistry, which wouldn't make sense for HSC). I've renamed these Handlers to make their use cases a lot clearer.